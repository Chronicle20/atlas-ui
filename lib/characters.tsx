import type {Tenant} from "@/types/models/tenant";
import type {UpdateCharacterData} from "@/types/models/character";
import {api} from "@/lib/api/client";

export interface Character {
    id: string;
    attributes: {
        accountId: number;
        worldId: number;
        name: string;
        level: number;
        experience: number;
        gachaponExperience: number;
        strength: number;
        dexterity: number;
        intelligence: number;
        luck: number;
        hp: number;
        maxHp: number;
        mp: number;
        maxMp: number;
        meso: number;
        hpMpUsed: number;
        jobId: number;
        skinColor: number;
        gender: number;
        fame: number;
        hair: number;
        face: number;
        mapId: number;
        spawnPoint: number;
        gm: number;
        x: number;
        y: number;
        stance: number;
    };
}

export async function fetchCharacters(tenant: Tenant): Promise<Character[]> {
    // Set tenant for this request
    api.setTenant(tenant);
    
    // Use the centralized API client to fetch characters
    return api.getList<Character>("/api/characters/");
}

export async function fetchCharacter(tenant: Tenant, characterId: string): Promise<Character> {
    // Set tenant for this request
    api.setTenant(tenant);
    
    // Use the centralized API client to fetch a single character
    return api.getOne<Character>(`/api/characters/${characterId}`);
}

export async function updateCharacter(tenant: Tenant, characterId: string, data: UpdateCharacterData): Promise<void> {
    // Set tenant for this request
    api.setTenant(tenant);
    
    // Prepare the JSON:API formatted request body
    const requestBody = {
        data: {
            type: "characters",
            id: characterId,
            attributes: data,
        },
    };
    
    // Use the centralized API client to update the character
    // The API client handles all error cases and status codes automatically
    await api.patch<void>(`/api/cos/characters/${characterId}`, requestBody);
}