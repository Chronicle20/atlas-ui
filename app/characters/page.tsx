"use client"

import {useTenant} from "@/context/tenant-context";
import {DataTable} from "@/components/data-table";
import {columns} from "@/app/characters/columns";
import {useEffect, useState} from "react";
import {Character, fetchCharacters} from "@/lib/characters";


export default function Page() {
    const {activeTenant} = useTenant();
    const [characters, setCharacters] = useState<Character[]>([]);
    const [loading, setLoading] = useState(true);
    const [error, setError] = useState<string | null>(null);

    useEffect(() => {
        if (!activeTenant) return;

        setLoading(true);

        fetchCharacters(activeTenant)
            .then((data) => {
                setCharacters(data);
            })
            .catch((err) => setError(err.message))
            .finally(() => setLoading(false));
    }, [activeTenant]);

    if (loading) return <div>Loading...</div>; // Show loading message while fetching data
    if (error) return <div>Error: {error}</div>; // Show error message if fetching failed

    return (
        <div className="flex flex-col flex-1 space-y-6 p-10 pb-16">
            <div className="items-center justify-between space-y-2">
                <div>
                    <h2 className="text-2xl font-bold tracking-tight">Characters</h2>
                </div>
            </div>
            <div className="mt-4">
                <DataTable columns={columns} data={characters}/>
            </div>
        </div>
    );
}