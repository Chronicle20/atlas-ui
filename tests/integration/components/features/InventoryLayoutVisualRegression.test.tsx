/**
 * Visual regression tests for inventory component layouts
 * Tests layout consistency, responsive behavior, and visual states
 */

import React from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { InventoryCard, InventoryCardSkeleton } from '@/components/features/characters/InventoryCard';
import { InventoryGrid } from '@/components/features/characters/InventoryGrid';
import type { Asset, Compartment } from '@/services/api/inventory.service';
import type { ItemDataResult } from '@/types/models/maplestory';

// Mock Next.js Image component for consistent visual testing
jest.mock('next/image', () => {
  return function MockImage({ src, alt, className, width, height, onLoad, onError, priority, unoptimized, ...props }: any) {
    // Simulate image loading for visual consistency
    React.useEffect(() => {
      const timer = setTimeout(() => {
        if (src && src.includes('valid-item')) {
          onLoad?.();
        } else if (src && src.includes('invalid-item')) {
          onError?.();
        }
      }, 10); // Fast loading for tests
      return () => clearTimeout(timer);
    }, [src, onLoad, onError]);

    return (
      <div
        data-testid="mock-image"
        data-src={src}
        data-alt={alt}
        data-width={width}
        data-height={height}
        data-priority={priority ? 'true' : 'false'}
        data-unoptimized={unoptimized ? 'true' : 'false'}
        style={{ width: width + 'px', height: height + 'px' }}
        className={className}
      />
    );
  };
});

// Mock errorLogger
jest.mock('@/services/errorLogger', () => ({
  errorLogger: {
    logError: jest.fn().mockResolvedValue(undefined),
  },
}));

// Mock hooks for controlled testing
const mockUseItemData = jest.fn();
const mockUseLazyLoad = jest.fn();
const mockUseItemDataCache = jest.fn();

jest.mock('@/lib/hooks/useItemData', () => ({
  useItemData: (...args: any[]) => mockUseItemData(...args),
  useItemDataCache: () => mockUseItemDataCache(),
}));

jest.mock('@/lib/hooks/useIntersectionObserver', () => ({
  useLazyLoad: (...args: any[]) => mockUseLazyLoad(...args),
}));

describe('Inventory Layout Visual Regression Tests', () => {
  let queryClient: QueryClient;

  beforeEach(() => {
    queryClient = new QueryClient({
      defaultOptions: {
        queries: {
          retry: false,
          gcTime: 0,
        },
      },
    });

    jest.clearAllMocks();
    
    // Default mocks
    mockUseLazyLoad.mockReturnValue({
      shouldLoad: true,
      ref: { current: null },
    });

    mockUseItemDataCache.mockReturnValue({
      warmCache: jest.fn().mockResolvedValue([]),
    });
  });

  afterEach(() => {
    queryClient.clear();
  });

  const createAsset = (templateId: number, slot: number): Asset => ({
    id: `asset-${templateId}-${slot}`,
    type: 'inventory',
    attributes: {
      characterId: 'test-char',
      slot,
      templateId,
      quantity: 1,
    },
  });

  const createCompartment = (capacity: number, type: number = 1): Compartment => ({
    id: `compartment-${type}`,
    type: 'compartment',
    attributes: {
      capacity,
      type,
    },
  });

  const createWrapper = () => {
    return ({ children }: { children: React.ReactNode }) => (
      <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
    );
  };

  describe('InventoryCard Layout Tests', () => {
    it('should maintain consistent dimensions across all states', async () => {
      const asset = createAsset(1302000, 0);
      
      // Test loading state
      mockUseItemData.mockReturnValueOnce({
        itemData: undefined,
        isLoading: true,
        hasError: false,
        errorMessage: undefined,
      });

      const { rerender, container } = render(
        <InventoryCard asset={asset} region="GMS" majorVersion={214} />,
        { wrapper: createWrapper() }
      );

      // Capture loading state dimensions
      const loadingCard = container.querySelector('.w-\\[100px\\].h-\\[120px\\]');
      expect(loadingCard).toBeInTheDocument();
      
      // Test loaded state with icon
      const mockItemData: ItemDataResult = {
        id: 1302000,
        name: 'Test Sword',
        iconUrl: 'https://maplestory.io/api/GMS/214/item/1302000/valid-item/icon',
        cached: false,
      };

      mockUseItemData.mockReturnValue({
        itemData: mockItemData,
        isLoading: false,
        hasError: false,
        errorMessage: undefined,
      });

      rerender(
        <InventoryCard asset={asset} region="GMS" majorVersion={214} />
      );

      await waitFor(() => {
        expect(screen.getByText('Test Sword')).toBeInTheDocument();
      });

      // Verify loaded state maintains same dimensions
      const loadedCard = container.querySelector('.w-\\[100px\\].h-\\[120px\\]');
      expect(loadedCard).toBeInTheDocument();

      // Test error state
      mockUseItemData.mockReturnValue({
        itemData: { id: 1302000, cached: false, error: 'Item not found' },
        isLoading: false,
        hasError: true,
        errorMessage: 'Item not found',
      });

      rerender(
        <InventoryCard asset={asset} region="GMS" majorVersion={214} />
      );

      // Verify error state maintains same dimensions
      const errorCard = container.querySelector('.w-\\[100px\\].h-\\[120px\\]');
      expect(errorCard).toBeInTheDocument();
      expect(screen.getByText('1302000')).toBeInTheDocument();
    });

    it('should render delete button consistently without layout shift', () => {
      const mockItemData: ItemDataResult = {
        id: 1302000,
        name: 'Test Item',
        iconUrl: 'https://maplestory.io/api/GMS/214/item/1302000/valid-item/icon',
        cached: false,
      };

      mockUseItemData.mockReturnValue({
        itemData: mockItemData,
        isLoading: false,
        hasError: false,
        errorMessage: undefined,
      });

      const asset = createAsset(1302000, 0);
      
      // Without delete button
      const { rerender, container } = render(
        <InventoryCard asset={asset} region="GMS" majorVersion={214} />,
        { wrapper: createWrapper() }
      );

      expect(container.querySelector('button')).not.toBeInTheDocument();

      // With delete button
      rerender(
        <InventoryCard 
          asset={asset} 
          region="GMS" 
          majorVersion={214} 
          onDelete={jest.fn()}
        />
      );

      const deleteButton = container.querySelector('button');
      expect(deleteButton).toBeInTheDocument();
      expect(deleteButton).toHaveClass('absolute', 'top-0', 'right-0');
      
      // Card dimensions should remain consistent
      const card = container.querySelector('.w-\\[100px\\].h-\\[120px\\]');
      expect(card).toBeInTheDocument();
    });

    it('should maintain consistent icon container dimensions', async () => {
      const asset = createAsset(1302000, 0);
      const mockItemData: ItemDataResult = {
        id: 1302000,
        name: 'Test Item',
        iconUrl: 'https://maplestory.io/api/GMS/214/item/1302000/valid-item/icon',
        cached: false,
      };

      mockUseItemData.mockReturnValue({
        itemData: mockItemData,
        isLoading: false,
        hasError: false,
        errorMessage: undefined,
      });

      render(
        <InventoryCard asset={asset} region="GMS" majorVersion={214} />,
        { wrapper: createWrapper() }
      );

      await waitFor(() => {
        expect(screen.getByText('Test Item')).toBeInTheDocument();
      });

      // Icon container should have fixed dimensions
      const iconContainer = screen.getByTestId('mock-image').parentElement;
      expect(iconContainer).toHaveClass('h-8', 'w-8');
    });

    it('should render text content within fixed height containers', () => {
      const asset = createAsset(1302000, 0);
      const mockItemData: ItemDataResult = {
        id: 1302000,
        name: 'Very Long Item Name That Should Wrap',
        iconUrl: 'https://maplestory.io/api/GMS/214/item/1302000/valid-item/icon',
        cached: false,
      };

      mockUseItemData.mockReturnValue({
        itemData: mockItemData,
        isLoading: false,
        hasError: false,
        errorMessage: undefined,
      });

      const { container } = render(
        <InventoryCard asset={asset} region="GMS" majorVersion={214} />,
        { wrapper: createWrapper() }
      );

      // Text container should have fixed height
      const textContainer = container.querySelector('.h-8.flex.items-center.justify-center');
      expect(textContainer).toBeInTheDocument();
      expect(textContainer).toHaveTextContent('Very Long Item Name That Should Wrap');
    });
  });

  describe('InventoryGrid Layout Tests', () => {
    it('should render consistent grid layout for different capacities', () => {
      const assets = [
        createAsset(1302000, 0),
        createAsset(1302001, 1),
        createAsset(1302002, 2),
      ];

      mockUseItemData.mockReturnValue({
        itemData: { id: 1302000, name: 'Test Item', cached: false },
        isLoading: false,
        hasError: false,
        errorMessage: undefined,
      });

      // Test small inventory (8 slots)
      const smallCompartment = createCompartment(8);
      const { rerender, container } = render(
        <InventoryGrid
          compartment={smallCompartment}
          assets={assets}
          region="GMS"
          majorVersion={214}
        />,
        { wrapper: createWrapper() }
      );

      expect(container.querySelector('.grid-cols-4')).toBeInTheDocument();
      expect(container.querySelectorAll('[data-testid^="inventory-slot-"]')).toHaveLength(8);

      // Test medium inventory (24 slots)
      const mediumCompartment = createCompartment(24);
      rerender(
        <InventoryGrid
          compartment={mediumCompartment}
          assets={assets}
          region="GMS"
          majorVersion={214}
        />
      );

      expect(container.querySelector('.grid-cols-6')).toBeInTheDocument();
      expect(container.querySelectorAll('[data-testid^="inventory-slot-"]')).toHaveLength(24);

      // Test large inventory (64 slots)
      const largeCompartment = createCompartment(64);
      rerender(
        <InventoryGrid
          compartment={largeCompartment}
          assets={assets}
          region="GMS"
          majorVersion={214}
        />
      );

      expect(container.querySelector('.grid-cols-8')).toBeInTheDocument();
      expect(container.querySelectorAll('[data-testid^="inventory-slot-"]')).toHaveLength(64);
    });

    it('should render empty slots with consistent styling', () => {
      const compartment = createCompartment(4);
      const assets: Asset[] = [createAsset(1302000, 1)]; // Only slot 1 filled

      mockUseItemData.mockReturnValue({
        itemData: { id: 1302000, name: 'Test Item', cached: false },
        isLoading: false,
        hasError: false,
        errorMessage: undefined,
      });

      const { container } = render(
        <InventoryGrid
          compartment={compartment}
          assets={assets}
          region="GMS"
          majorVersion={214}
        />,
        { wrapper: createWrapper() }
      );

      // Should have 4 slots total
      const slots = container.querySelectorAll('[data-testid^="inventory-slot-"]');
      expect(slots).toHaveLength(4);

      // Empty slots should have consistent styling
      const emptySlots = container.querySelectorAll('.border-dashed');
      expect(emptySlots).toHaveLength(3); // 3 empty slots

      emptySlots.forEach(slot => {
        expect(slot).toHaveClass('w-[100px]', 'h-[120px]', 'border-2', 'border-dashed');
      });
    });

    it('should maintain grid gap consistency', () => {
      const compartment = createCompartment(12);
      const assets = Array.from({ length: 6 }, (_, i) => createAsset(1302000 + i, i));

      mockUseItemData.mockReturnValue({
        itemData: { id: 1302000, name: 'Test Item', cached: false },
        isLoading: false,
        hasError: false,
        errorMessage: undefined,
      });

      const { container } = render(
        <InventoryGrid
          compartment={compartment}
          assets={assets}
          region="GMS"
          majorVersion={214}
        />,
        { wrapper: createWrapper() }
      );

      // Grid should have consistent gap
      const grid = container.querySelector('.grid');
      expect(grid).toHaveClass('gap-3');
    });

    it('should render loading state with skeleton cards', () => {
      const compartment = createCompartment(8);
      const assets: Asset[] = [];

      const { container } = render(
        <InventoryGrid
          compartment={compartment}
          assets={assets}
          region="GMS"
          majorVersion={214}
          isLoading={true}
        />,
        { wrapper: createWrapper() }
      );

      // Should render skeleton cards (limited to 8 for this compartment)
      const skeletons = container.querySelectorAll('.w-\\[100px\\].h-\\[120px\\]');
      expect(skeletons.length).toBeGreaterThan(0);
      expect(skeletons.length).toBeLessThanOrEqual(8);
    });
  });

  describe('Responsive Layout Behavior', () => {
    it('should handle different compartment types with optimal layouts', () => {
      const assets = [createAsset(1302000, 0)];

      mockUseItemData.mockReturnValue({
        itemData: { id: 1302000, name: 'Test Item', cached: false },
        isLoading: false,
        hasError: false,
        errorMessage: undefined,
      });

      // Test EQUIPABLES compartment (type 1)
      const equipCompartment = createCompartment(16, 1);
      const { rerender, container } = render(
        <InventoryGrid
          compartment={equipCompartment}
          assets={assets}
          region="GMS"
          majorVersion={214}
        />,
        { wrapper: createWrapper() }
      );

      expect(container.querySelector('.grid-cols-4')).toBeInTheDocument();

      // Test ETC compartment (type 4)
      const etcCompartment = createCompartment(32, 4);
      rerender(
        <InventoryGrid
          compartment={etcCompartment}
          assets={assets}
          region="GMS"
          majorVersion={214}
        />
      );

      expect(container.querySelector('.grid-cols-8')).toBeInTheDocument();
    });

    it('should maintain card aspect ratios across all screen sizes', () => {
      const asset = createAsset(1302000, 0);
      const mockItemData: ItemDataResult = {
        id: 1302000,
        name: 'Test Item',
        iconUrl: 'https://maplestory.io/api/GMS/214/item/1302000/valid-item/icon',
        cached: false,
      };

      mockUseItemData.mockReturnValue({
        itemData: mockItemData,
        isLoading: false,
        hasError: false,
        errorMessage: undefined,
      });

      const { container } = render(
        <InventoryCard asset={asset} region="GMS" majorVersion={214} />,
        { wrapper: createWrapper() }
      );

      // Card should maintain fixed dimensions regardless of content
      const card = container.querySelector('.w-\\[100px\\].h-\\[120px\\]');
      expect(card).toBeInTheDocument();

      // Aspect ratio should be approximately 5:6 (100:120)
      // This ensures consistent visual layout across different devices
    });
  });

  describe('Layout State Transitions', () => {
    it('should prevent layout shift during loading to loaded transition', async () => {
      const asset = createAsset(1302000, 0);

      // Start with loading state
      mockUseItemData.mockReturnValueOnce({
        itemData: undefined,
        isLoading: true,
        hasError: false,
        errorMessage: undefined,
      });

      const { rerender, container } = render(
        <InventoryCard asset={asset} region="GMS" majorVersion={214} />,
        { wrapper: createWrapper() }
      );

      // Capture initial skeleton layout
      const initialSkeleton = screen.getByTestId('inventory-card-loading');
      expect(initialSkeleton).toBeInTheDocument();

      // Transition to loaded state
      const mockItemData: ItemDataResult = {
        id: 1302000,
        name: 'Loaded Item',
        iconUrl: 'https://maplestory.io/api/GMS/214/item/1302000/valid-item/icon',
        cached: false,
      };

      mockUseItemData.mockReturnValue({
        itemData: mockItemData,
        isLoading: false,
        hasError: false,
        errorMessage: undefined,
      });

      rerender(
        <InventoryCard asset={asset} region="GMS" majorVersion={214} />
      );

      await waitFor(() => {
        expect(screen.getByText('Loaded Item')).toBeInTheDocument();
      });

      // Card dimensions should remain consistent
      const card = container.querySelector('.w-\\[100px\\].h-\\[120px\\]');
      expect(card).toBeInTheDocument();
    });

    it('should maintain visual consistency during error state transition', async () => {
      const asset = createAsset(1302000, 0);

      // Start with loaded state
      mockUseItemData.mockReturnValueOnce({
        itemData: {
          id: 1302000,
          name: 'Test Item',
          iconUrl: 'https://maplestory.io/api/GMS/214/item/1302000/valid-item/icon',
          cached: false,
        },
        isLoading: false,
        hasError: false,
        errorMessage: undefined,
      });

      const { rerender, container } = render(
        <InventoryCard asset={asset} region="GMS" majorVersion={214} />,
        { wrapper: createWrapper() }
      );

      await waitFor(() => {
        expect(screen.getByText('Test Item')).toBeInTheDocument();
      });

      // Transition to error state
      mockUseItemData.mockReturnValue({
        itemData: { id: 1302000, cached: false, error: 'Item not found' },
        isLoading: false,
        hasError: true,
        errorMessage: 'Item not found',
      });

      rerender(
        <InventoryCard asset={asset} region="GMS" majorVersion={214} />
      );

      // Should show fallback with template ID
      expect(screen.getByText('1302000')).toBeInTheDocument();
      expect(screen.getByTestId('inventory-card-package-icon')).toBeInTheDocument();

      // Card dimensions should remain consistent
      const card = container.querySelector('.w-\\[100px\\].h-\\[120px\\]');
      expect(card).toBeInTheDocument();
    });
  });

  describe('Visual Layout Snapshots', () => {
    it('should match snapshot for loaded inventory card', async () => {
      const asset = createAsset(1302000, -11);
      const mockItemData: ItemDataResult = {
        id: 1302000,
        name: 'Basic Sword',
        iconUrl: 'https://maplestory.io/api/GMS/214/item/1302000/valid-item/icon',
        cached: false,
      };

      mockUseItemData.mockReturnValue({
        itemData: mockItemData,
        isLoading: false,
        hasError: false,
        errorMessage: undefined,
      });

      const { container } = render(
        <InventoryCard asset={asset} region="GMS" majorVersion={214} />,
        { wrapper: createWrapper() }
      );

      await waitFor(() => {
        expect(screen.getByText('Basic Sword')).toBeInTheDocument();
      });

      // Snapshot test for visual consistency
      expect(container.firstChild).toMatchSnapshot('loaded-inventory-card');
    });

    it('should match snapshot for inventory grid with mixed content', async () => {
      const compartment = createCompartment(8);
      const assets = [
        createAsset(1302000, 0),
        createAsset(1302001, 2),
        createAsset(1302002, 5),
      ];

      mockUseItemData.mockReturnValue({
        itemData: { id: 1302000, name: 'Grid Item', cached: false },
        isLoading: false,
        hasError: false,
        errorMessage: undefined,
      });

      const { container } = render(
        <InventoryGrid
          compartment={compartment}
          assets={assets}
          region="GMS"
          majorVersion={214}
        />,
        { wrapper: createWrapper() }
      );

      // Wait for content to render
      await waitFor(() => {
        const renderedAssets = screen.getAllByText('Grid Item');
        expect(renderedAssets).toHaveLength(3);
      });

      // Snapshot test for grid layout consistency
      expect(container.firstChild).toMatchSnapshot('inventory-grid-mixed-content');
    });
  });
});