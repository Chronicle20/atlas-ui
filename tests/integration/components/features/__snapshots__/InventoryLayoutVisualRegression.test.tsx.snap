// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Inventory Layout Visual Regression Tests Visual Layout Snapshots should match snapshot for inventory grid with mixed content: inventory-grid-mixed-content 1`] = `
<div
  class="grid gap-3 grid-cols-4"
>
  <div
    class="flex items-center justify-center"
    data-slot="0"
    data-testid="inventory-slot-0"
  >
    <div
      class="bg-card text-card-foreground flex flex-col gap-6 rounded-xl border shadow-sm overflow-hidden relative py-0 w-[100px] h-[120px]"
      data-slot="card"
    >
      <div
        class="@container/card-header grid auto-rows-min grid-rows-[auto_auto] items-start gap-1.5 has-data-[slot=card-action]:grid-cols-[1fr_auto] [.border-b]:pb-6 p-1 pl-2 pb-1 text-left"
        data-slot="card-header"
      >
        <div
          class="text-xs font-medium text-muted-foreground"
        >
          0
        </div>
      </div>
      <div
        class="p-2 pt-0 flex flex-col items-center justify-center min-h-[70px]"
        data-slot="card-content"
      >
        <div
          class="flex flex-col items-center space-y-1"
        >
          <div
            class="relative h-8 w-8 flex-shrink-0"
          >
            <div
              class="flex items-center justify-center h-full w-full"
            >
              <svg
                aria-hidden="true"
                class="lucide lucide-package h-6 w-6 text-muted-foreground flex-shrink-0"
                data-testid="inventory-card-package-icon"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M11 21.73a2 2 0 0 0 2 0l7-4A2 2 0 0 0 21 16V8a2 2 0 0 0-1-1.73l-7-4a2 2 0 0 0-2 0l-7 4A2 2 0 0 0 3 8v8a2 2 0 0 0 1 1.73z"
                />
                <path
                  d="M12 22V12"
                />
                <polyline
                  points="3.29 7 12 12 20.71 7"
                />
                <path
                  d="m7.5 4.27 9 5.15"
                />
              </svg>
            </div>
          </div>
          <div
            class="text-xs text-center break-all h-8 flex items-center justify-center w-full"
          >
            <span
              class="font-medium line-clamp-2 leading-tight"
            >
              Grid Item
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="flex items-center justify-center"
    data-slot="1"
    data-testid="inventory-slot-1"
  >
    <div
      class="w-[100px] h-[120px] border-2 border-dashed border-muted-foreground/30 rounded-lg flex flex-col items-center justify-center bg-muted/10 hover:bg-muted/20 transition-colors"
    >
      <div
        class="text-xs font-medium text-muted-foreground/60 mb-2"
      >
        1
      </div>
      <div
        class="text-xs text-muted-foreground/50"
      >
        Empty
      </div>
    </div>
  </div>
  <div
    class="flex items-center justify-center"
    data-slot="2"
    data-testid="inventory-slot-2"
  >
    <div
      class="bg-card text-card-foreground flex flex-col gap-6 rounded-xl border shadow-sm overflow-hidden relative py-0 w-[100px] h-[120px]"
      data-slot="card"
    >
      <div
        class="@container/card-header grid auto-rows-min grid-rows-[auto_auto] items-start gap-1.5 has-data-[slot=card-action]:grid-cols-[1fr_auto] [.border-b]:pb-6 p-1 pl-2 pb-1 text-left"
        data-slot="card-header"
      >
        <div
          class="text-xs font-medium text-muted-foreground"
        >
          2
        </div>
      </div>
      <div
        class="p-2 pt-0 flex flex-col items-center justify-center min-h-[70px]"
        data-slot="card-content"
      >
        <div
          class="flex flex-col items-center space-y-1"
        >
          <div
            class="relative h-8 w-8 flex-shrink-0"
          >
            <div
              class="flex items-center justify-center h-full w-full"
            >
              <svg
                aria-hidden="true"
                class="lucide lucide-package h-6 w-6 text-muted-foreground flex-shrink-0"
                data-testid="inventory-card-package-icon"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M11 21.73a2 2 0 0 0 2 0l7-4A2 2 0 0 0 21 16V8a2 2 0 0 0-1-1.73l-7-4a2 2 0 0 0-2 0l-7 4A2 2 0 0 0 3 8v8a2 2 0 0 0 1 1.73z"
                />
                <path
                  d="M12 22V12"
                />
                <polyline
                  points="3.29 7 12 12 20.71 7"
                />
                <path
                  d="m7.5 4.27 9 5.15"
                />
              </svg>
            </div>
          </div>
          <div
            class="text-xs text-center break-all h-8 flex items-center justify-center w-full"
          >
            <span
              class="font-medium line-clamp-2 leading-tight"
            >
              Grid Item
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="flex items-center justify-center"
    data-slot="3"
    data-testid="inventory-slot-3"
  >
    <div
      class="w-[100px] h-[120px] border-2 border-dashed border-muted-foreground/30 rounded-lg flex flex-col items-center justify-center bg-muted/10 hover:bg-muted/20 transition-colors"
    >
      <div
        class="text-xs font-medium text-muted-foreground/60 mb-2"
      >
        3
      </div>
      <div
        class="text-xs text-muted-foreground/50"
      >
        Empty
      </div>
    </div>
  </div>
  <div
    class="flex items-center justify-center"
    data-slot="4"
    data-testid="inventory-slot-4"
  >
    <div
      class="w-[100px] h-[120px] border-2 border-dashed border-muted-foreground/30 rounded-lg flex flex-col items-center justify-center bg-muted/10 hover:bg-muted/20 transition-colors"
    >
      <div
        class="text-xs font-medium text-muted-foreground/60 mb-2"
      >
        4
      </div>
      <div
        class="text-xs text-muted-foreground/50"
      >
        Empty
      </div>
    </div>
  </div>
  <div
    class="flex items-center justify-center"
    data-slot="5"
    data-testid="inventory-slot-5"
  >
    <div
      class="bg-card text-card-foreground flex flex-col gap-6 rounded-xl border shadow-sm overflow-hidden relative py-0 w-[100px] h-[120px]"
      data-slot="card"
    >
      <div
        class="@container/card-header grid auto-rows-min grid-rows-[auto_auto] items-start gap-1.5 has-data-[slot=card-action]:grid-cols-[1fr_auto] [.border-b]:pb-6 p-1 pl-2 pb-1 text-left"
        data-slot="card-header"
      >
        <div
          class="text-xs font-medium text-muted-foreground"
        >
          5
        </div>
      </div>
      <div
        class="p-2 pt-0 flex flex-col items-center justify-center min-h-[70px]"
        data-slot="card-content"
      >
        <div
          class="flex flex-col items-center space-y-1"
        >
          <div
            class="relative h-8 w-8 flex-shrink-0"
          >
            <div
              class="flex items-center justify-center h-full w-full"
            >
              <svg
                aria-hidden="true"
                class="lucide lucide-package h-6 w-6 text-muted-foreground flex-shrink-0"
                data-testid="inventory-card-package-icon"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M11 21.73a2 2 0 0 0 2 0l7-4A2 2 0 0 0 21 16V8a2 2 0 0 0-1-1.73l-7-4a2 2 0 0 0-2 0l-7 4A2 2 0 0 0 3 8v8a2 2 0 0 0 1 1.73z"
                />
                <path
                  d="M12 22V12"
                />
                <polyline
                  points="3.29 7 12 12 20.71 7"
                />
                <path
                  d="m7.5 4.27 9 5.15"
                />
              </svg>
            </div>
          </div>
          <div
            class="text-xs text-center break-all h-8 flex items-center justify-center w-full"
          >
            <span
              class="font-medium line-clamp-2 leading-tight"
            >
              Grid Item
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="flex items-center justify-center"
    data-slot="6"
    data-testid="inventory-slot-6"
  >
    <div
      class="w-[100px] h-[120px] border-2 border-dashed border-muted-foreground/30 rounded-lg flex flex-col items-center justify-center bg-muted/10 hover:bg-muted/20 transition-colors"
    >
      <div
        class="text-xs font-medium text-muted-foreground/60 mb-2"
      >
        6
      </div>
      <div
        class="text-xs text-muted-foreground/50"
      >
        Empty
      </div>
    </div>
  </div>
  <div
    class="flex items-center justify-center"
    data-slot="7"
    data-testid="inventory-slot-7"
  >
    <div
      class="w-[100px] h-[120px] border-2 border-dashed border-muted-foreground/30 rounded-lg flex flex-col items-center justify-center bg-muted/10 hover:bg-muted/20 transition-colors"
    >
      <div
        class="text-xs font-medium text-muted-foreground/60 mb-2"
      >
        7
      </div>
      <div
        class="text-xs text-muted-foreground/50"
      >
        Empty
      </div>
    </div>
  </div>
</div>
`;

exports[`Inventory Layout Visual Regression Tests Visual Layout Snapshots should match snapshot for loaded inventory card: loaded-inventory-card 1`] = `
<div
  class="bg-card text-card-foreground flex flex-col gap-6 rounded-xl border shadow-sm overflow-hidden relative py-0 w-[100px] h-[120px]"
  data-slot="card"
>
  <div
    class="@container/card-header grid auto-rows-min grid-rows-[auto_auto] items-start gap-1.5 has-data-[slot=card-action]:grid-cols-[1fr_auto] [.border-b]:pb-6 p-1 pl-2 pb-1 text-left"
    data-slot="card-header"
  >
    <div
      class="text-xs font-medium text-muted-foreground"
    >
      -11
    </div>
  </div>
  <div
    class="p-2 pt-0 flex flex-col items-center justify-center min-h-[70px]"
    data-slot="card-content"
  >
    <div
      class="flex flex-col items-center space-y-1"
    >
      <div
        class="relative h-8 w-8 flex-shrink-0"
      >
        <div
          class="object-contain transition-opacity duration-200 opacity-0"
          data-alt="Basic Sword"
          data-height="32"
          data-priority="true"
          data-src="https://maplestory.io/api/GMS/214/item/1302000/valid-item/icon"
          data-testid="mock-image"
          data-unoptimized="true"
          data-width="32"
          style="width: 32px; height: 32px;"
        />
        <div
          class="bg-muted animate-pulse absolute inset-0 h-8 w-8 rounded"
        />
      </div>
      <div
        class="text-xs text-center break-all h-8 flex items-center justify-center w-full"
      >
        <span
          class="font-medium line-clamp-2 leading-tight"
        >
          Basic Sword
        </span>
      </div>
    </div>
  </div>
</div>
`;
